File: Leetcode/Python3/LinkedList/partition.py
# -*- coding:UTF-8 -*-
'''
Solution:
- 设置一个dummy node
- 设置一个lessNode，用来记录当前小于x的区域
- 用cur去遍历list，同时记录pre。遇到小于x的，pre直接指向cur的next，lessNode指向cur即可；
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        
        lessNode = dummy
        pre, cur = dummy, head
        while cur:
            if cur.val < x:
                if pre == lessNode:
                    lessNode = lessNode.next
                    tmp = cur.next
                    pre = cur
                    cur = tmp
                else:
                    pre.next = cur.next
                    tmp = lessNode.next
                    lessNode.next = cur
                    cur.next = tmp
                    
                    lessNode = lessNode.next
                    cur = pre.next
            else:
                tmp = cur.next
                pre = cur
                cur = tmp
                
        return dummy.next

File: Leetcode/C++/TwoPointer/partition.cpp
/*
Solution: 时间复杂度O(n),空间复杂度O(n)
- 另外定义两个链表，分别存放小于x的node和>=x的node；
- 注意p2->next = NULL，因为p2是原来链表中的结点，后面可能还有其他结点，不置空会导致返回的新链表成环；
- 这里注意点操作符(.)和箭头操作符(->)的使用，.左边必须是实体，->左边必须是指针，具体看https://blog.csdn.net/qq457163027/article/details/54237782.

Solution 2：时间复杂度O(n),空间复杂度O(1) 有待完成（TODO）。
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode dummy1(0), dummy2(0);
        ListNode *p1 = &dummy1, *p2 = &dummy2;
        while(head){
            if(head->val < x){
                p1->next = head;
                p1 = p1->next;
            }
            else{
                p2->next = head;
                p2 = p2->next;
            }
            head = head->next;
        }
        p1->next = dummy2.next;
        p2->next = NULL;
        return dummy1.next;
        
    }
};